### ISSUE TITLE
Support for psycopg3

### ISSUE THREAD

**Issue by CaselIT** [sqlalchemy/sqlalchemy#6842]
### Describe the use case

SQLAlchemy will need to add support for psycopg3 at some point.

From the article [here](https://www.psycopg.org/articles/2021/08/02/psycopg3-django-driver/) it seems that its behaviour will be similar to asyncpg wrt execution, so it may require adding casting to the queries and/or rewrite some cases.
I think care must be taken to avoid cases similar to https://github.com/sqlalchemy/sqlalchemy/issues/6833

I'm not sure if it still supports execution without server side parameterized query.

We may also need to decide how to call the dialect, since it seems that the project is called psycopg3 but the package is imported as `psycopg`

### Databases / Backends / Drivers targeted

PostrgreSQL

### Example Use

Not relevant

### Additional context

Current wip patch: https://github.com/sqlalchemy/sqlalchemy/issues/6842#issuecomment-919531590

**Comment by dvarrazzo** [comment]
Hello, psycopg developer here. Please let us know if you need a hand on our side :)

**Comment by CaselIT** [comment]
Hi, @dvarrazzo. Thanks!

Just a question, the pypi package is https://pypi.org/project/psycopg3 or https://pypi.org/project/psycopg ?

**Comment by dvarrazzo** [comment]
It is the latter. The release is not visible, because it's a beta. https://pypi.org/project/psycopg/3.0b1/

Original idea was to released a package called psycopg3, but that makes semver harder to apply. So there will be a psycopg4 sometimes and that won't be a drama :smile:

**Comment by dvarrazzo** [comment]
FYI, looking at #6833, I don't think psycopg 3 has problem with `to_tsvector`:

```python
In [1]: import psycopg

In [2]: conn = psycopg.connect()

In [3]: conn.execute("select to_tsvector(%s, %s)", ("english", "test")).fetchone()
Out[3]: ("'test':1",)
```

There would be a problem if sqlachemy adds a cast, but from a couple of tests I'm doing it seems that asyncpg has a much more simplified adaptation system:

```
>>> await (await asyncpg.connect()).fetch("select $1 + $2", 1, 1)
...
AmbiguousFunctionError: operator is not unique: unknown + unknown
HINT:  Could not choose a best candidate operator. You might need to add explicit type casts.
```

I think psycopg works better without casts in several places:

```
>>> psycopg.connect().execute("select %s + %s", (1, 1)).fetchone()
(2,)
```

**Comment by CaselIT** [comment]
That's good to ear. Mine was just a supposition due to the server side parameters binding.

**Comment by CaselIT** [comment]
@dvarrazzo deos the async version behaves the same?

**Comment by dvarrazzo** [comment]
Sync and async behave the same, yes. They only differ on what they do when they have to wait.

**Comment by CaselIT** [comment]
What should the dialect be called?
I guess `psycopg3` within sqlalchemy would still makes sense, even if the package it's called psycopg.

What's your preference @dvarrazzo @zzzeek ?

**Comment by zzzeek** [comment]
psycopg3 for sure.  if "psycopg" does a version 4 we will likely need a psycopg4 also if the API changes that much

**Comment by zzzeek** [comment]
@dvarrazzo  it will be "`import psycopg`" ?   i sort of like 'import psycopg3' better but either way

**Comment by CaselIT** [comment]
> 
> 
> @dvarrazzo it will be "`import psycopg`" ? i sort of like 'import psycopg3' better but either way

seems so: https://www.psycopg.org/psycopg3/docs/basic/usage.html

```py
# Note: the module name is psycopg, not psycopg3
import psycopg

# Connect to an existing database
with psycopg.connect("dbname=test user=postgres") as conn:
  ...
```

**Comment by dvarrazzo** [comment]
I cannot foresee the future. Psycopg 2 was first released 15 years ago, so this is really a major rewrite. Just sticking to ordinary semver, Psycopg 4 could be released in one year time for relatively minor, but non-backward compatible changes.

Maybe calling it psycopg3 would be more future-proof. What policy do you follow with other adapters?

**Comment by Kludex** [comment]
`beautifulsoup` does something interesting (like an "alias"): https://pypi.org/project/beautifulsoup4/ and https://pypi.org/project/bs4/. jfyk

**Comment by CaselIT** [comment]
> 
> 
> `beautifulsoup` does something interesting (like an "alias"): https://pypi.org/project/beautifulsoup4/ and https://pypi.org/project/bs4/. jfyk

could you then run `import bs4` ?

**Comment by Kludex** [comment]
> > `beautifulsoup` does something interesting (like an "alias"): https://pypi.org/project/beautifulsoup4/ and https://pypi.org/project/bs4/. jfyk
> 
> could you then run `import bs4` ?

Exactly.

**Comment by zzzeek** [comment]
note that SQLAlchemy is doing "sqlalchemy 2" but we are calling it "sqlalchemy", because trying to run both would be a disaster.   for psycopg, the name is changing regardless so from a technical point of view it likely doesnt matter.   psycopg3 feels more natural since we've been typing a number for 15 years but im not yet thinking  of other reasons besides that.

**Comment by dvarrazzo** [comment]
My suggestion would be "psycopg" just because the idea is that psycopg > 3, in the future, would be still psycopg, then connecting using `postgresql+psycopg3` to psycopg4 would be kind of weird... Of course you can introduce new version numbers in the dialect names.

psycopg3 felt natural for me for a long time too...

**Comment by CaselIT** [comment]
No dbapi (from memory) have a version number in the name, so I think if we decide for a version it would stay the same until the `import` changes

I guess that if the plan is to follow semver, `psycopg` may make more sense than `psycopg3` since it may be strange to have v4 of that, Unless there is a clean distinction between the name (`psycopg3`) and the version, like starting from v1 to make it clear that the 3 is not the package version. Or calver may also be used.

On the other hand it seems that psycopg2 > psycopg to an user that is just looking for a way of connecting to a db.

**Comment by CaselIT** [comment]
@dvarrazzo I was looking at the isolation level in psycopg3.

Autocommit is now separate from the isolation_level. Does it mean that if I want to set autocommit should I do
```py
connection.autocommit = True
connection.isolation_level = None
```
and vice versa to set another level should I do
```py
connection.autocommit = False
connection.isolation_level = IsolationLevel.REPEATABLE_READ
```

Or setting one automatically adjust the other?
From the documentation is not clear what interaction the two values attribute they have https://www.psycopg.org/psycopg3/docs/api/connections.html?highlight=autocommit#psycopg.Connection.autocommit

In psycopg2 it was all managed by [set_isolation_level](https://www.psycopg.org/docs/connection.html?highlight=set_isolation_level#connection.set_isolation_level)

**Comment by sqla-tester** [comment]
**Federico Caselli** has proposed a fix for this issue in the **master** branch:

**Added support for ``psycopg3`` dialect.** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3081

**Comment by CaselIT** [comment]
just to get the ball rolling. still lots of work to do

**Comment by dvarrazzo** [comment]
> @dvarrazzo I was looking at the isolation level in psycopg3.
> 
> Autocommit is now separate from the isolation_level. Does it mean that if I want to set autocommit should I do
> 
> ```python
> connection.autocommit = True
> connection.isolation_level = None
> ```

Hi @CaselIT No, I don't think so. Setting autocommit to true will make psycopg not emit the BEGIN to start a transaction. If an isolation level is chosen, then the transaction initial statement might be BEGIN ISOLATION LEVEL SERIALIZABLE, but autocommit mode there the isolation level won't have effect, unless people then start a transaction explicitly using `conn.transaction()`.

So I would say set autocommit = True and leave the isolation level unchanged.

**Comment by CaselIT** [comment]
Ok, while in the opposite case autocommit should be set to false when setting an isolation level.

Basically my goal is to make it behave like the https://www.psycopg.org/docs/extensions.html#psycopg2.extensions.ISOLATION_LEVEL_AUTOCOMMIT of psycopg2

**Comment by dvarrazzo** [comment]
I am not sure I would try to replicate the semantic of psycopg 2. Is there a reason for doing that?

Personally, if people sets autocommit, I would just set autocommit, not touch the isolation level. And similarly wouldn't change autcommit if someone changes isolation level.

**Comment by CaselIT** [comment]
That's how sqlalchemy works. Autocommit is treated as an isolation level like in psycopg2.
https://docs.sqlalchemy.org/en/14/core/connections.html#setting-transaction-isolation-levels-including-dbapi-autocommit

Not sure if it's feasible to change that. Cc @zzzeek

**Comment by CaselIT** [comment]
Is there a way of turning off the numeric adaptation https://www.psycopg.org/psycopg3/docs/basic/adapt.html#numbers-adaptation at least for smallint?

I get errors with some functions, like `generate_series`:
```
function generate_series(smallint, smallint) is not unique
LINE 2: FROM generate_series($1, $2) AS x JOIN generate_series($3, $...
```

**Comment by CaselIT** [comment]
@dvarrazzo I think there is a bug in `register_hstore`:

```py
import psycopg
from psycopg.types import TypeInfo
from psycopg.types.hstore import register_hstore

conn = psycopg.connect("postgresql://scott:tiger@127.0.0.1:5432/test")
info = TypeInfo.fetch(conn, "hstore")
register_hstore(info)

conn.execute('CREATE TABLE foo(id integer, bar hstore[])')
conn.execute(
    "INSERT INTO foo (id, bar) VALUES (%(id)s, %(bar)s::HSTORE[])",
    {"id": 1, "bar": [{"a": "1"}, {"b": "1"}]},
)
print(conn.execute('select * from foo').fetchall())
conn.rollback()
```

calling `register_hstore(info)` to register it globally does not work for this use case, and raises `psycopg.ProgrammingError: cannot adapt type dict to format AUTO`.
If I instead register it on the connection using `register_hstore(info, conn)`, it works and prints: `[(1, [{'a': '1'}, {'b': '1'}])]`

**Comment by dvarrazzo** [comment]
@CaselIT 

> @dvarrazzo I think there is a bug in `register_hstore`:

no, it's working as expected. The types registry and adapter maps are copy-on-write. Once a connection or a cursor is created, changing a wider context doesn't affect the objects already created.

**Comment by dvarrazzo** [comment]
> Is there a way of turning off the numeric adaptation https://www.psycopg.org/psycopg3/docs/basic/adapt.html#numbers-adaptation at least for smallint?

I see the problem: generate_series() has types integer, bigint, numeric, but no smallint. No, there isn't a way to disable cast to smallint: it would have bad side effects (e.g. failing to insert in a smallint column without forcing a cast).

You should use a cast to a supported data type such as `generate_series(%s::int, %s::int)` or wrap the arguments to one of the wrappers in `psycopg.types.numeric` such as `Int4`, `Int8`, `IntNumeric`.

**Comment by CaselIT** [comment]
Thanks for the support!

> 
> 
> @CaselIT
> 
> > @dvarrazzo I think there is a bug in `register_hstore`:
> 
> no, it's working as expected. The types registry and adapter maps are copy-on-write. Once a connection or a cursor is created, changing a wider context doesn't affect the objects already created.

Thanks. The issue presented since I'm registering this globally on initialization after the first connection is done, so it did not propagate to the existing connection. I guess I can just register in both or make sqlalchemy discard this first connection.

May I suggest mentioning it in the documentation of the functions that take a context (I guess also `set_json_dumps` and other behave the same) that globally does not apply to existing object? Something like 
> The context where to <...>. If None, register it globally for newly created contexts (connections, cursors, etc).

---

> 
> 
> > Is there a way of turning off the numeric adaptation https://www.psycopg.org/psycopg3/docs/basic/adapt.html#numbers-adaptation at least for smallint?
> 
> I see the problem: generate_series() has types integer, bigint, numeric, but no smallint. No, there isn't a way to disable cast to smallint: it would have bad side effects (e.g. failing to insert in a smallint column without forcing a cast).
> 
> You should use a cast to a supported data type such as `generate_series(%s::int, %s::int)` or wrap the arguments to one of the wrappers in `psycopg.types.numeric` such as `Int4`, `Int8`, `IntNumeric`.

Thanks for the clarification, I'll probably use explicit casts for the `generate_series`. The `row` function is another case that seems to require it. I'm trying to avoid having to cast everything like in asyncpg

**Comment by dvarrazzo** [comment]
> May I suggest mentioning it in the documentation of the functions that take a context (I guess also set_json_dumps and other behave the same) that globally does not apply to existing object?

Yes, you are right, and I understand the confusion. Will document that better.

> I'll probably use explicit casts for the generate_series. The row function is another case that seems to require it. I'm trying to avoid having to cast everything like in asyncpg

Hopefully there shouldn't be as many cases as in asyncpg: I have tried to tune the adaptation system to do the right thing in most cases. Funny that it trips on `generate_series()`: it is used a lot in the test suite. Looking at it, I have always used it with only one parameter in the test suite. This fails:

```
In [4]: cnn.execute("select generate_series(%s, %s)", (1,4)).fetchall()
AmbiguousFunction: function generate_series(smallint, smallint) is not unique
```
but this works alright:
```
In [8]: cnn.execute("select generate_series(1, %s)", (4,)).fetchall()
Out[8]: [(1,), (2,), (3,), (4,)]
```

**Comment by CaselIT** [comment]
Postgress is strange at times :)

Btw I've yet to read the documentation, but on some test that create and destroy enums I get "type xxxxx not found in the cache" or something similar. What would be the correct way of invalidating the type cache of a connection? IIRC we had to do something similar also for asyncpg

**Comment by dvarrazzo** [comment]
> What would be the correct way of invalidating the type cache of a connection?

In all honesty, never thought about that. Can you explain better, or propose a way to reproduce the issue? Thank you

**Comment by CaselIT** [comment]
Ok the tests do something like this:
```py
import psycopg

conn = psycopg.connect("postgresql://scott:tiger@127.0.0.1:5432/test")

for i in range(3):
    conn.execute("""CREATE TYPE an_enum AS ENUM ('foo', 'bar', 'baz');""")
    conn.execute("CREATE TABLE foo(id integer, bar an_enum[])")
    conn.cursor().executemany(
        "INSERT INTO foo (bar) VALUES (%(enum_col)s::an_enum[])",
        [{"enum_col": ["foo"]}, {"enum_col": ["foo", "bar"]}],
    )
    print(conn.execute("select * from foo").fetchall())
    conn.rollback()
```

On the second try this error is raised by the executemany:
> psycopg.errors.InternalError_: cache lookup failed for type 29065

**Comment by CaselIT** [comment]
Note that the array is not needed for the exception to raise

**Comment by CaselIT** [comment]
In the test there is also a similar error:
> psycopg.errors.FeatureNotSupported: cached plan must not change result type

That happens in the a query like this: ` SELECT foo.bar FROM foo ORDER BY foo.id`.

I think this is connected to the above one, meaning that we have parametrized test running so it re-creates the table with different types, causing the error.

We could probably just invalidate the connections between tests, so that a new one is created before use each time, since I believe a similar usage is only present in test scenarios.

**Comment by dvarrazzo** [comment]
Uhm... that exception comes from the server. It can be reproduced in psql:

```
BEGIN;
CREATE TYPE an_enum AS ENUM ('foo', 'bar', 'baz');
CREATE TABLE foo(id integer, bar an_enum[]);
PREPARE stmt (an_enum[]) AS INSERT INTO foo (bar) VALUES ($1);
EXECUTE stmt ('{foo}');
ROLLBACK;

BEGIN;
CREATE TYPE an_enum AS ENUM ('foo', 'bar', 'baz');
CREATE TABLE foo(id integer, bar an_enum[]);
EXECUTE stmt ('{foo}');
ERROR:  cache lookup failed for type 8072147

\errverbose 
ERROR:  XX000: cache lookup failed for type 8072147
LOCATION:  getBaseTypeAndTypmod, lsyscache.c:2355
```

It's caused by psycopg using prepared statements in executemany, but it might be caused by normal execution too because statements get prepared after seen a few times.

I'll have a chat with the Postgres people to understand if we are doing something wrong on our side.

**Comment by dvarrazzo** [comment]
Smells like it would be good to discard prepared statements on rollback, but I'd like to hear from the postgres people if that's the expected behaviour (with the error to become explicit rather than internal)

**Comment by CaselIT** [comment]
For reference in asyncpg we call https://magicstack.github.io/asyncpg/current/api/index.html?highlight=reload_schema_state#asyncpg.connection.Connection.reload_schema_state

**Comment by dvarrazzo** [comment]
I think that asyncpg deals with the schema in a more implicit way, resulting in the need of invalidating its cache. In psycopg the steps to interact with the schema are explicit so it is sort of obvious that they should be repeated (e.g. if `TypeInfo.fetch()` is called in a transaction, with an object created and then discarded in that transaction, it is obvious that the same info won't work in a new transaction).

I will think if it's possible to do something less brutal than discarding all the connection prepared statements on rollback.

**Comment by CaselIT** [comment]
Just to confirm, currently the only way of fixing this is to reconnect, right?

**Comment by dvarrazzo** [comment]
@CaselIT I have forwarded the issue to the Postgres devs. [Their answer](https://www.postgresql.org/message-id/2220999.1632335261%40sss.pgh.pa.us), as predictable is "don't do that" :slightly_smiling_face: 

I have pushed a changeset which should clear the prepared statement caches: this means:

- on ROLLBACK
- on ROLLBACK TO SAVEPOINT
- on any DROP * statement

This solves the problem you reported, see the tests added in https://github.com/psycopg/psycopg/commit/7b11ddc90203a52812ee0099c32c7b1c7b9992d6

Can you test that this changeset fixes your issues? If so I will merge it to master, otherwise we can maybe improve it.

Thank you!

**Comment by CaselIT** [comment]
> Can you test that this changeset fixes your issues? If so I will merge it to master, otherwise we can maybe improve it.

It does seem to work. I no longer get `cache lookup failed for type 29065` or  `cached plan must not change result type` type of errors.

Thanks for the quick turnaround!

**Comment by CaselIT** [comment]
I'm not sure if this is psycopg or it's postgres that's interpreting it like this:
```py
import psycopg


conn = psycopg.connect("postgresql://scott:tiger@127.0.0.1:5432/test")


conn.execute("CREATE TABLE foo(id integer, bar integer[])")
conn.execute(
    "DELETE FROM foo WHERE foo.bar = %(bar_1)s", {"bar_1": [1, 2, 3, 4, 5]}
)
```
fails with `operator does not exist: integer[] = numeric[]`

**Comment by dvarrazzo** [comment]
> It does seem to work. 

Great! I'll merge it then. Thank you for the report and for testing!

**Comment by dvarrazzo** [comment]
> I'm not sure if this is psycopg or it's postgres that's interpreting it like this:

It's psycopg I'm afraid. When given a list of Python integers it's hard to guess what data type to use (we should find the largest number in absolute value). So psycopg uses a data type that can fit them all. The rationale for doing that was that postgres is able to cast down anyway:

```
*piro=# create table foo(ints integer[]);
CREATE TABLE
*piro=# insert into foo values ('{1}'::numeric[]);
INSERT 0 1
```

I didn't expect that comparison fail

```
piro=# select '{1}'::integer[] = '{1}'::numeric[];
ERROR:  operator does not exist: integer[] = numeric[]
```

Probably that's because the cast is "AS ASSIGNMENT"  rather than "AS IMPLICIT".

I don't think it could be worked around even if we went O(n) over the list to find the min/max. In your example we would have picked `smallint[]` and failed with the same operator not existing error. Other strategy would be to dump the array as unknown but that only works in text mode and has other drawbacks.

I am not sure we can do better than that on our side. Can you do anything on yours? It would be to add a cast on the rhs there, which I think it's only necessary for `integer[]`, `smallint[]`, `bigint[]`.

**Comment by CaselIT** [comment]
> I am not sure we can do better than that on our side. Can you do anything on yours? It would be to add a cast on the rhs there, which I think it's only necessary for `integer[]`, `smallint[]`, `bigint[]`.

I think we can do something here. I'll start by adding a cast to all arrays, then go from there. We can probably relax that a bit if there's a need. From experience this is required also for array of custom types, like enum (I think we need to add that also in psycopg2)

Btw, there is still some "psycopg 3" wording in these docs https://www.psycopg.org/psycopg3/docs/basic/install.html

**Comment by dvarrazzo** [comment]
> From experience this is required also for array of custom types, like enum (I think we need to add that also in psycopg2)

Not necessarily, I've made quite some work to make the adaptation system behave. Enums should work ok. Feel free to try and let us know if something doesn't work as expected.

> Btw, there is still some "psycopg 3" wording in these docs https://www.psycopg.org/psycopg3/docs/basic/install.html

The important is that there is no `psycopg3` :) but I'll review for consistency.

**Comment by CaselIT** [comment]
> 
> 
> > From experience this is required also for array of custom types, like enum (I think we need to add that also in psycopg2)
> 
> Not necessarily, I've made quite some work to make the adaptation system behave. Enums should work ok. Feel free to try and let us know if something doesn't work as expected.

Ok, that's something to try. I think I'll first try making it pass all the test, then work on removing unnecessary things

> 
> > Btw, there is still some "psycopg 3" wording in these docs https://www.psycopg.org/psycopg3/docs/basic/install.html
> 
> The important is that there is no `psycopg3` :) but I'll review for consistency.

I have not seen it.

I've a question, more like an use case, maybe it's better as an issue in psycopg tracker.

In the dialect I wanted to register hstore and user provided json dumps/loads globally when requiresed provided, but I think it's not advisable, since it causes issues both in the tests suite that necessarily try different combinations, and because doing so would mean that an user cannot have two engines with different configurations. So I think I'll move the registration per-connection, like in the other dialects.

Looking at how the registration happens in psycopg, my question/use case was: would it be possible for psycopg to have a "Context" class concept that could be used to register these adapters. Then in the connect call I could pass the context and the connection inherits from it? The idea being that each dialect (so basically each engine) would have its own context, I would do the registration only once, not on each connection. Something like this:

```py
ctx = psycopg.context()
psycopg.types.json.set_json_loads(my_loads, ctx)
psycopg.types.hstore.register_hstore(hstore_info, ctx)
# other register stuff with ctx
conn = psycopg.connect(..., context=ctx)
```

Note that none of this is blocking or even required, just an use case that I though of while working on the dialect, and maybe also useful for direct users of psycopg, like a library that uses it could have its own customization in a context, avoiding polluting the global state.

If you prefer it as an issue in the psycopg tracker, I can file it there.

**Comment by dvarrazzo** [comment]
> would it be possible for psycopg to have a "Context" class concept that could be used to register these adapters

I have thought about that, when writing the django driver, because the situation is similar, with [several adapters registered](https://github.com/dvarrazzo/django-psycopg3-backend/blob/acc326023e4305148d327c3b500f7cab4c45bd5d/django_psycopg3/base.py#L208) for every connection. Other use cases might be the use of psycopg with other non-quite-standard postgres, such as with CockroachDb or Amazon RDS. That's the reason why the default map is in  `psycopg.postgres.adapters`.

Using a pre-configured context has also a slight performance benefit because customising it for every connection creates a deep copy, whereas passing a template to the connection makes use of the copy-on-write and leaves the connection copy shallow.

So yes, maybe open a ticket on our tracker, but the idea is there and the implementation is as simple as making a context reach [`BaseConnection.__init__`](https://github.com/psycopg/psycopg/blob/d8be1ceb68e4a1794acc44e94444dda685c62544/psycopg/psycopg/connection.py#L107).

**Comment by CaselIT** [comment]
Thanks for considering it. I'll open an issue here https://github.com/psycopg/psycopg/issues

**Comment by dvarrazzo** [comment]
Ok, in the meantime [this branch]()https://github.com/psycopg/psycopg/compare/connection-custom-context] should do what you expect...

Please give it a testing and if it works I'll merge to master.

**Comment by CaselIT** [comment]
Nice!
So basically I just do this:
```
ctx = psycopg.adapt.AdaptersMap()
psycopg.types.json.set_json_loads(my_loads, ctx)
psycopg.types.hstore.register_hstore(hstore_info, ctx)
# other register stuff with ctx
conn = psycopg.connect(..., context=ctx)
```
Thanks for the quick turnaround, I'll try it later.

**Comment by dvarrazzo** [comment]
Pretty much, except that that this would leave you with an empty context, apart from what you put there explicitly. If you want to start with the default Postgres configuration you can initialise `ctx = psycopg.adapt.AdaptersMap(psycopg.adapters)`.

**Comment by CaselIT** [comment]
> 
> 
> Ok, in the meantime [this branch](https://github.com/psycopg/psycopg/compare/connection-custom-context) should do what you expect...
> 
> Please give it a testing and if it works I'll merge to master.

It works great for this use case!

**Comment by CaselIT** [comment]
I've updated https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3081

The sync version should be in a good state. Missing:
- server side cursors,
- execute many features of psycopg2 (not sure if it's supported)
- recheck the skips since some of them require updating (they target only psycopg2)
- docs

To test it needs to install psycopg from this branch https://github.com/psycopg/psycopg/compare/connection-custom-context

then run with `pytest --db psycopg3 ./test/dialect/postgresql/`

Regarding the async version I wonder if we can keep the same driver name. I'll take a look

**Comment by dvarrazzo** [comment]
> execute many features of psycopg2 (not sure if it's supported)

the "fast execution helpers"? No, they are not. executemany is already much more performing and copy is much easier to use.

> To test it needs to install psycopg from this branch https://github.com/psycopg/psycopg/compare/connection-custom-context

I'll give a check to the changes to the django backend and merge to master.

**Comment by zzzeek** [comment]
we are planning to generalize "fast executemany helpers" to all postgresql drivers and maybe even other backends.   If you have an INSERT statement in Postgresql, running multiple rows in multiple VALUES clauses is lightning fast plus we get RETURNING for it, so we definitely want to use that.   asyncpg's "executemany" is slow as well without this, are you sure psycopg3's native "executemany" is faster than the "fast execution helpers" ?

**Comment by dvarrazzo** [comment]
No, I haven't benchmarked the VALUES solution against executemany and I can see how it can be faster. However these execution helpers are simple to implement. I think psycopg2 solution is far from being the only possible or the best solution.

Psycopg 3 adds a python-based copy support that would be probably even faster than that, although with no returning.

Are you planning to have a common SA interface to several drivers and construct the query and the values in SA?

**Comment by zzzeek** [comment]
More or less yes, we've already implemented the framework for the helpers in the compiler.  There's a GitHub issue here I don't have the number handy that lays out the plan to generalize it into sqlalchemy core.

**Comment by CaselIT** [comment]
@dvarrazzo Another query that requires casting, then the parameters is None:
```sql
SELECT date_table.id
FROM date_table
WHERE CASE WHEN ($1 IS NOT NULL) THEN $1 ELSE date_table.date_data END = date_table.date_data
```
where `date_data` is any time like datatype (date, timestamp, time)

I can work around it by adding a cast when sqlalchemy knows the type and the value is None, but that would break caching, right @zzzeek ?

**Comment by zzzeek** [comment]
As long as the structure of the query generates the same sql every time without any dependency on the structure of literal data presented to it at execution time,  there's no issue with caching.

**Comment by zzzeek** [comment]
For example we generate different sql for "x == 1" vs "x == None", that's fine you just can't use a bound parameter there and pass None at execution time.

**Comment by CaselIT** [comment]
I was handling it in `visit_bindparam`, meaning that it would result in `"x == 1"` vs ` "x == None::integer"`.

But in any case that would be very similar to always casting, since in most cases bindparams that are not literal will have a value of None

**Comment by zzzeek** [comment]
right you can't do it there.  you have to receive the None and then produce cast() or whatever based on that value, *before* the cache key is generated.  that is, the fact that None vs. non-None was passed has to affect the SQL construct that is created such that it will have a different cache key.

**Comment by dvarrazzo** [comment]
> > To test it needs to install psycopg from this branch https://github.com/psycopg/psycopg/compare/connection-custom-context
> 
> I'll give a check to the changes to the django backend and merge to master.

Heads up: I have merged this branch to master.

There's another change in master since beta 1 which might affect you: I have [dropped `connection.client_encoding`](https://www.psycopg.org/psycopg3/docs/basic/from_pg2.html#client-encoding-is-gone), because I don't think it is of much use any more. There are still workarounds to set it, it just doesn't need a prominent attribute on the connection.

**Comment by CaselIT** [comment]
Thanks for the heads up. I'll update to support `client_encoding`.

Have you had a change at looking if the query reported in https://github.com/sqlalchemy/sqlalchemy/issues/6842#issuecomment-926917070 can avoid using casts? I'm not sure if that's pg fault or something that psycopg can help with

**Comment by dvarrazzo** [comment]
If you rewrite the expression as `WHERE coalesce($1, date_table.date_data) = date_table.date_data` it doesn't need a cast, because the type can be inferred from the context.

OTOH, if we associated an arbitrary type to `None`, Postgres would refuse to perform most of the casts.

It makes me think that there isn't really a way to specify a NULL with type associated from psycopg 3: None always goes down untyped, so a cast on the placeholder seems the only way. I don't have in mind a way to add this feature though which wouldn't be pretty invasive (from the point of the end user).

**Comment by CaselIT** [comment]
I've update the first message to have a link to the patch.

It still does not support some edge cases wrt cases and we have they to look at the execute many optimizations.
Other than that is should work.

Please share any you have feedback once you have tried it, thanks!

**Comment by CaselIT** [comment]
oh nice, [psycopg](https://pypi.org/project/psycopg/) is now out of beta

**Comment by CaselIT** [comment]
>  The latter would raise even more confusion when there will be a version 4 of psycopg.

well one is the name of the library the other is the version, I don't necessarily see the connection between library name and library version. But I agree that it may seem strange

**Comment by sqla-tester** [comment]
**Federico Caselli** has proposed a fix for this issue in the **main** branch:

**Added support for ``psycopg`` dialect.** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3201

**Comment by CaselIT** [comment]
@dvarrazzo The changeset at gerrit is mortly ready, it's just missing some documentation. If you can have a look it would be great! https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3081

Thanks!

**Comment by dvarrazzo** [comment]
I had a look, and it seems mostly ok from my PoV, just left a few comments.

@zzzeek opened https://github.com/psycopg/psycopg/issues/157 which I need to look into better.

**Comment by CaselIT** [comment]
> I had a look, and it seems mostly ok from my PoV, just left a few comments.

thanks. the fetch row after connection closure seems to continue to happen, maybe that's because that test makes changes the connection in status "BAD" and not "CLOSED"? See gerrit for details

**Comment by jacopofar** [comment]
If I understand correctly, this is not expected to be available before the 2.0 release, right?

**Comment by CaselIT** [comment]
Yes, this change refactored quite a bit of the current psycopg2 dialect also so it was deemed safer to have this on v2

also @zzzeek it seems that we have lost the changelog entry somehow. I think it was left in the 1.4 folder and it's now no longer in either folder.

**Comment by zzzeek** [comment]
ok that's bad because that would mean its in the release notes , at leat in the 2.0 docs

**Comment by zzzeek** [comment]
got removed in this merge here: 3b6a68bc15b5c44b25d21f66

**Comment by CaselIT** [comment]
does not seem to be in the release notes.

btw I was pushing the same change also.

**Comment by AntonOfTheWoods** [comment]
@CaselIT , because 2.0 has been waiting now for years, might it be clearer to reopen this issue so that people don't look at the issue and see it as "closed" because support is already available?

**Comment by CaselIT** [comment]
I'm -1 on reopening this since by that logic any issue that's targeting v2 should be left open, and that's not really feasible

We may have been too prudent in this case not backporting it to the 1.4 release, but that ship has already sailed.

**Comment by zzzeek** [comment]
has there been some announcement  / change regarding psycopg3 that users are suddenly looking for this library?

**Comment by AntonOfTheWoods** [comment]
I am using both psycopg2 and asyncpg and have been wanting to migrate to 3 to just use one lib since they went RTM. I was doing some spring cleaning and noticed it again today, and noticed this has been closed for almost a year.

I certainly appreciate all the complexities and apologise if I was appearing ungrateful!

**Comment by zzzeek** [comment]
sorry I am grumpy, I want 2.0 to be released as much as anyone, the final two commits are in review now then I have to get beta 1 ready to go.  so this is happening now

**Comment by stemid** [comment]
Seeing as this is closed, what do we subscribe to in order to track progress on psycopg3 support?

**Comment by CaselIT** [comment]
version 2 was released 6 months ago

**Comment by stemid** [comment]
> version 2 was released 6 months ago

yeah sorry I deleted my comment because I immediately realized I had to write postgresql+psycopg in the URI to use psycopg3, otherwise it kept using psycopg2.
